### PR TITLE
Fix sampling error with many pipelines when allow_repeats=False

### PR DESCRIPTION
**Pull request by caciolai**
**What does this PR do? Please describe:**
A summary of the change or the issue that is fixed.

Fixes https://github.com/facebookresearch/fairseq2/issues/1472

**Does your PR introduce any breaking changes? If yes, please list them:**
List of all backwards-incompatible changes.

**Check list:**
- [x] Was the content of this PR **discussed and approved** via a GitHub issue? (no need for typos or documentation improvements)
- [x] Did you read the [contributor guideline](https://github.com/facebookresearch/fairseq2/blob/main/CONTRIBUTING.md)?
- [x] Did you make sure that your **PR does only one thing** instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**?
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [ ] Did you **update the [CHANGELOG](https://github.com/facebookresearch/fairseq2/blob/main/CHANGELOG.md)**? (no need for typos, documentation, or minor internal changes)

### CODE REVIEW

File: native/src/fairseq2n/data/sample_data_source.cc [hunk 1]
```diff
@@ -212,9 +237,10 @@ sample_data_source::block(std::size_t idx)
         weight_cumsums_[i] -= weight;
     }
 
-    float32 sum = weight_cumsums_.back();
+    // Find the actual maximum cumulative weight (total remaining probability mass)
+    float32 sum = *std::max_element(weight_cumsums_.begin(), weight_cumsums_.end());
```
Comment (caciolai) [inline native/src/fairseq2n/data/sample_data_source.cc]: this might not be needed

File: native/src/fairseq2n/data/sample_data_source.cc [hunk 2]
```diff
@@ -212,9 +237,10 @@ sample_data_source::block(std::size_t idx)
...
 
-    if (!are_close(sum, 1.0F)) {
+    if (!are_close(sum, 1.0F) && sum > 0.0F) {
```
Comment (caciolai) [inline native/src/fairseq2n/data/sample_data_source.cc]: this might not be needed

File: native/src/fairseq2n/data/sample_data_source.cc [hunk 3]
```diff
@@ -170,9 +175,26 @@ sample_data_source::random_pipeline_index()
...
 
+    // The binary search returns the index where cumsum[idx] >= sample and cumsum[idx-1] < sample
+    // If this pipeline is blocked, it means cumsum[idx] == cumsum[idx-1], which is a
+    // degenerate interval. We need to search forward to find the pipeline that actually
+    // owns this probability mass (the next unblocked one with cumsum[idx] > cumsum[idx-1])
+    while (lptr < weight_cumsums_.size() && is_blocked_[lptr]) {
+        lptr++;
+    }
+
+    // If we went past the end, wrap around and search from the beginning
```
Comment (caciolai) [inline native/src/fairseq2n/data/sample_data_source.cc]: this should not be needed

File: native/src/fairseq2n/data/sample_data_source.cc [hunk 4]
```diff
@@ -20,14 +20,11 @@
...
 
-sample_data_source::sample_data_source(
-    std::vector<data_pipeline> &&pipelines,
-    std::vector<float32> &&weights,
-    std::optional<std::uint64_t> maybe_seed,
-    bool allow_repeats)
-  : pipelines_(std::move(pipelines)), 
-    is_epoch_done_(pipelines_.size()),
-    allow_repeats_{allow_repeats}
+sample_data_source::sample_data_source(std::vector<data_pipeline> &&pipelines,
```
Comment (cbalioglu) [inline native/src/fairseq2n/data/sample_data_source.cc]: Do you mind keeping the formatting consistent with the rest of the code?

File: tests/unit/data/data_pipeline/test_sample.py [hunk 5]
```diff
@@ -253,3 +254,23 @@ def test_op_saves_and_restores_state_when_allow_repeats_false_is_specified(
             assert list(pipeline) == [3, 4, 7, 8, 9, 10, 11, 12]
 
             pipeline.reset(reset_rng=True)
+
+    def test_op_works_with_many_pipelines_when_allow_repeats_false(self) -> None:
+        """Test edge case with many pipelines where only lowest-weight ones remain."""
+
+        # See https://github.com/fairinternal/fairseq2-ext/issues/181
```
Comment (cbalioglu) [inline tests/unit/data/data_pipeline/test_sample.py]: Please remove any internal references in OSS code.

File: tests/unit/data/data_pipeline/test_sample.py [hunk 6]
```diff
@@ -253,3 +254,23 @@ def test_op_saves_and_restores_state_when_allow_repeats_false_is_specified(
...
+            pipelines.append(read_sequence(list(range(s))).and_return())
+
+        builder = DataPipeline.sample(pipelines, weights, seed=123, allow_repeats=False)
+        pipeline = builder.and_return()
+
+        expected_total_size = np.sum(sizes)
+        actual_sampled_size = len([_ for _ in iter(pipeline)])
+        assert (
+            actual_sampled_size == expected_total_size
+        ), f"Expected {expected_total_size}, got {actual_sampled_size}"
```
Comment (cbalioglu) [inline tests/unit/data/data_pipeline/test_sample.py]: The assertion text is not really needed since this is a pytest test.

File: tests/unit/data/data_pipeline/test_sample.py [hunk 7]
```diff
@@ -253,3 +254,23 @@ def test_op_saves_and_restores_state_when_allow_repeats_false_is_specified(
...
+        sizes = np.random.RandomState(0).randint(0, 100_000, nb)
+        pipelines = []
+        for s in sizes:
+            pipelines.append(read_sequence(list(range(s))).and_return())
+
+        builder = DataPipeline.sample(pipelines, weights, seed=123, allow_repeats=False)
+        pipeline = builder.and_return()
+
+        expected_total_size = np.sum(sizes)
+        actual_sampled_size = len([_ for _ in iter(pipeline)])
```
Comment (cbalioglu) [inline tests/unit/data/data_pipeline/test_sample.py]: Nit: `len(list(pipeline))`

File: native/src/fairseq2n/data/sample_data_source.cc [hunk 8]
```diff
@@ -170,9 +175,16 @@ sample_data_source::random_pipeline_index()
             rptr = mptr;
     }
 
+    // The binary search returns the index where cumsum[idx] >= sample and cumsum[idx-1] < sample
+    // If this pipeline is blocked, it means cumsum[idx] == cumsum[idx-1], which is a
+    // degenerate interval. We need to search forward to find the pipeline that actually
+    // owns this probability mass (the next unblocked one with cumsum[idx] > cumsum[idx-1])
+    while (lptr < weight_cumsums_.size() && is_blocked_[lptr]) {
```
Comment (cbalioglu) [inline native/src/fairseq2n/data/sample_data_source.cc]: As discussed offline, this can be reduced to:

```python

if (!allow_repeats_) {
  while (is_epoch_done_[lptr])
    lptr++;
}
```

I would also put `lptr < weight_cumsums_.size()` optionally as an assert statement since due to `are_all_done()` check, `lptr` should never move beyond the range of `is_epoch_done_`.

File: native/src/fairseq2n/data/sample_data_source.cc [hunk 9]
```diff
@@ -170,6 +170,17 @@ sample_data_source::random_pipeline_index()
...
+    // Binary search finds an index where cumsum[idx] >= sample.
+    // When allow_repeats_ is false, exhausted pipelines have cumsum[idx] == cumsum[idx-1].
+    // Skip forward to find the next active pipeline that "owns" this probability mass.
+    if (!allow_repeats_) { 
+        while (lptr < is_epoch_done_.size() && is_epoch_done_[lptr]) lptr++;
+    }
+
+    // Should never happen since at least one pipeline is active 
+    // (guaranteed by are_all_done() check in next())
+    assert(lptr < weight_cumsums_.size());
```
Comment (cbalioglu) [inline native/src/fairseq2n/data/sample_data_source.cc]: I suggest moving this within the loop and removing the `lptr < ..` check from the loop condition. We likely won't reach this assert anyways since `is_epoch_done_[lptr]` will cause a failure (e.g. SEGFAULT) while looping.

File: doc/source/reference/fairseq2.data_type.rst [hunk 10]
```diff
@@ -8,5 +8,4 @@ fairseq2.data_type
 Functions
 =========
 
-.. autofunction:: current_dtype
```
Comment (cbalioglu) [inline doc/source/reference/fairseq2.data_type.rst]: This is a "bug" on my end. Let me fix it in another PR and rebase this PR on latest main afterwards.

File: native/src/fairseq2n/data/sample_data_source.cc (hunk 9 above)
Comment (caciolai) [reply native/src/fairseq2n/data/sample_data_source.cc]: like so?
```

        while (is_epoch_done_[lptr]) {
            lptr++;
            // Should never happen since at least one pipeline is active 
            // (guaranteed by are_all_done() check in next())
            assert(lptr < is_epoch_done_.size());
        }
```
just for my curiosity, what's the difference? is it "cosmetic" or does it have practical ramifications?

Comment (cbalioglu) [reply native/src/fairseq2n/data/sample_data_source.cc]: It has a semantic difference. If you include the `lptr < ..` check within the loop condition, it implies that this is an expected condition. When you put it as an assert statement within the loop, you make it clear that this is an invariant and should never happen unless the implementation has a bug in it. So instead of silently moving out of the loop when that condition is met, we stop the process with an assertion error.

Comment (caciolai) [reply native/src/fairseq2n/data/sample_data_source.cc]: Thanks for explaining!

Comment (cbalioglu) [reply native/src/fairseq2n/data/sample_data_source.cc]: No problem!
